pkg/reconciler/ingress/resources/virtual_service.go

```diff
@@ -153,7 +154,11 @@ func makeVirtualServiceSpec(ing *v1alpha1.Ingress, gateways map[v1alpha1.Ingress
 func makeVirtualServiceRoute(hosts sets.String, http *v1alpha1.HTTPIngressPath, gateways map[v1alpha1.IngressVisibility]sets.String, visibility v1alpha1.IngressVisibility) *istiov1alpha3.HTTPRoute {
 	matches := []*istiov1alpha3.HTTPMatchRequest{}
 	clusterDomainName := network.GetClusterDomainName()
-	for _, host := range hosts.List() {
+
+	// Deduplicate hosts to avoid excessive matches, which cause a combinatorial expansion in Istio
+	distinctHosts := getDistinctHostPrefixes(hosts)
+
+	for _, host := range distinctHosts.List() {
 		g := gateways[visibility]
 		if strings.HasSuffix(host, clusterDomainName) && len(gateways[v1alpha1.IngressVisibilityClusterLocal]) > 0 {
```

Comment:
One thing I am not sure of, which is causing tests to fail: This suffix check is failing now, since we are deduplicating the full hostname out. However, I don't see how it ever worked before.

For example, if I have a KIngress with `hosts: [a.default, a.default.svc, a.default.svc.cluster.local]; visibility: ExternalIP`, then we would decide that only the last host should be ClusterLocal, when in reality they are all aliases for the same host so they should probably all have the same visibility.

I would consider this a bug, and remove this detection and assume that rules have the proper visibility set. However, I am not sure the implications of this, so I left it as is. This causes the tests to fail so some advice is needed.

Reply:
Thanks! I updated the tests to accommodate for this new behavior